core/capabilities/syncer.go

```diff
@@ -56,26 +71,486 @@ var defaultStreamConfig = p2ptypes.StreamConfig{
+	// TODO: Remove this when we're no longer hard-coding
+	// a `networkSetup`.
+	for p, cfg := range s.networkSetup.allPeers {
+		allPeers[p] = cfg
+	}
+
+	publicDONs := []kcr.CapabilityRegistryDONInfo{}
+	for _, d := range readerState.DONs {
+		if !d.IsPublic {
+			continue
+		}
+
+		publicDONs = append(publicDONs, d)
+
+		for _, nid := range d.NodeP2PIds {
+			allPeers[nid] = defaultStreamConfig
+		}
+	}
+	err = s.peerWrapper.GetPeer().UpdateConnections(allPeers)
+	if err != nil {
+		return fmt.Errorf("failed to update peer connections: %w", err)
+	}
+
+	// Next, we need to split the DONs into the following:
+	// - workflow DONs the current node is a part of.
+	// These will need remote shims to all remote capabilities on other DONs.
+	//
+	// We'll also construct a set to record what DONs the current node is a part of,
+	// regardless of any modifiers (public/acceptsWorkflows etc).
+	myID := s.peerWrapper.GetPeer().ID()
+	myWorkflowDONs := []kcr.CapabilityRegistryDONInfo{}
+	remoteWorkflowDONs := []kcr.CapabilityRegistryDONInfo{}
+	myDONs := map[uint32]bool{}
+	for _, d := range readerState.DONs {
+		for _, peerID := range d.NodeP2PIds {
+			if peerID == myID {
+				myDONs[d.Id] = true
+			}
+		}
+
+		if d.AcceptsWorkflows {
+			if myDONs[d.Id] {
+				myWorkflowDONs = append(myWorkflowDONs, d)
+			} else {
+				remoteWorkflowDONs = append(remoteWorkflowDONs, d)
+			}
+		}
+	}
+
+	// - remote capability DONs (with IsPublic = true) the current node is a part of.
+	// These need server-side shims.
+	myCapabilityDONs := []kcr.CapabilityRegistryDONInfo{}
+	remoteCapabilityDONs := []kcr.CapabilityRegistryDONInfo{}
+	for _, d := range publicDONs {
+		if len(d.CapabilityConfigurations) > 0 {
+			if myDONs[d.Id] {
+				myCapabilityDONs = append(myCapabilityDONs, d)
+			} else {
+				remoteCapabilityDONs = append(remoteCapabilityDONs, d)
+			}
+		}
+	}
+
+	// Now, if my node is a workflow DON, let's setup any shims
+	// to external capabilities.
+	if len(myWorkflowDONs) > 0 {
+		myDON := myWorkflowDONs[0]
+
+		// TODO: this is a bit nasty; figure out how best to handle this.
+		if len(myWorkflowDONs) > 1 {
+			s.lggr.Warn("node is part of more than one workflow DON; assigning first DON as caller")
+		}
+
+		for _, rcd := range remoteCapabilityDONs {
+			err := s.addRemoteCapabilities(ctx, myDON, rcd, readerState)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Finally, if I'm a capability DON, let's enable external access
+	// to the capability.
+	if len(myCapabilityDONs) > 0 {
+		for _, mcd := range myCapabilityDONs {
+			err := s.enableExternalAccess(ctx, myID, mcd, readerState, remoteWorkflowDONs)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func signersFor(don kcr.CapabilityRegistryDONInfo, state state) ([][]byte, error) {
+	s := [][]byte{}
+	for _, nodeID := range don.NodeP2PIds {
+		node, ok := state.IDsToNodes[nodeID]
+		if !ok {
+			return nil, fmt.Errorf("could not find node for id %s", nodeID)
+		}
```

Comment:
External registry contract enforces this, so this check is redundant.

Reply:
This should be `criticial` if it does happen 🤷‍♂️

Reply:
That's why we're bubbling up the error here; I'd expect this never to happen but it's here so that we can flag invariant violations like this up the stack where it will get logged.